config/deploy/docker/lib/aws_sdk_helpers.rb

```diff
@@ -0,0 +1,128 @@
+    define_method(:ecr)              { AwsSdkHelpers::ClientMethods.ecr }
+    define_method(:ecs)              { AwsSdkHelpers::ClientMethods.ecs }
+    define_method(:secretsmanager)   { AwsSdkHelpers::ClientMethods.secretsmanager }
+    define_method(:autoscaling)      { AwsSdkHelpers::ClientMethods.autoscaling }
+    define_method(:ec2)              { AwsSdkHelpers::ClientMethods.ec2 }
+    define_method(:cloudwatchevents) { AwsSdkHelpers::ClientMethods.cloudwatchevents }
+    define_method(:cw)               { AwsSdkHelpers::ClientMethods.cw }
+    define_method(:cwl)              { AwsSdkHelpers::ClientMethods.cwl }
+  end
+
+  module Helpers
+    include AwsSdkHelpers::ClientMethods
+
+    def self.cluster_name
+      ENV.fetch('CLUSTER_NAME', 'openpath')
+    end
+    def _cluster_name
+      AwsSdkHelpers::Helpers.cluster_name
+    end
+
+    def self.capacity_providers(cluster)
+      r = {}
+      capacity_provider_names = AwsSdkHelpers::ClientMethods.ecs.describe_clusters(clusters: [cluster]).clusters.first.capacity_providers
+      AwsSdkHelpers::ClientMethods.ecs.describe_capacity_providers(capacity_providers: capacity_provider_names).capacity_providers.each do |capacity_provider|
+        asg_name = capacity_provider.auto_scaling_group_provider.auto_scaling_group_arn.split('/').last
+        asg = AwsSdkHelpers::ClientMethods.autoscaling.describe_auto_scaling_groups(auto_scaling_group_names: [ asg_name ]).auto_scaling_groups.first
+
+        launch_template_id = asg.mixed_instances_policy.launch_template.launch_template_specification.launch_template_id
+        launch_template_versions = AwsSdkHelpers::ClientMethods.ec2.describe_launch_template_versions(launch_template_id: launch_template_id)
+
+        ami_id = launch_template_versions.launch_template_versions[0].launch_template_data.image_id
+
+        r[capacity_provider.name] = {
+          name: capacity_provider.name,
+          ami_id: ami_id
+        }
+      end
+      return r
+    end
+    def _capacity_providers(cluster = nil)
+      cluster ||= _cluster_name
+      @capacity_providers ||= AwsSdkHelpers::Helpers.capacity_providers(cluster)
+    end
+
+    def self.default_placement_constraints(ami_id:, cluster_name:, capacity_provider_name:)
+      # Confirm that there is at least one instance w/the provided AMI ID.
+      payload = {
+        filters: [
+          {
+            name: "tag:ecs",
+            values: ["true"],
+          },
+          {
+            name: "tag:Cluster",
+            values: [cluster_name],
+          },
+          {
+            name: "tag:capacity-provider.name",
+            values: [capacity_provider_name]
+          },
+          {
+            name: "image-id",
+            values: [ami_id]
+          }
+        ]
+      }
+      matching_instances = ClientMethods.ec2.describe_instances(payload)
+
+      if matching_instances.reservations.count > 0 && !ami_id.empty?
```

Comment:
This is better, but let's talk it through before we merge this. I still have this feeling we can end up with un-schedulable containers.

If you start deploying after the AMI change, the capacity provider might have enough capacity and not spin up a new instance, and if you start using instance refresh in the ASG, you're draining all the services, getting one new-ami instance at a time, but the services still want the old AMI.

Could we update the constraints as part of an instance refresh (little by little)?

My current thinking is that 1 new ami instance might not be enough and that we'll need a process to artificially expand the cluster temporarily (like with a bunch of nginx containers with medium sized RAM quotas). Each deployment after an update would retire some of the nginx containers.

I don't think this is the best idea, so let's talk to the group. Maybe I'm overthinking this.

Might have to take a step back and use a new capacity provider whenever we have a new AMI. That might simplify all of this. I think we might be finally seeing why they don't recommend constraints with capacity providers.

Reply:
That sounds right.